### PR TITLE
unsafe recovery: Introduce auto-detect mode for online recovery

### PR DESCRIPTION
**Pull request by Connor1996**
Signed-off-by: Connor1996 <zbk602423539@gmail.com>

### What problem does this PR solve?

Issue Number: Close https://github.com/tikv/pd/issues/5415

### What is changed and how does it work?

```commit-message
Support auto-detect failed stores for online recovery
```

### Check List

Tests

- Unit test

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):

### Release note

```release-note
Support auto-detect failed stores for online recovery
```

### CODE REVIEW

**Review by v01dstar — APPROVED** [review]
LGTM

File: server/api/unsafe_operation.go [hunk 1]
```diff
@@ -48,24 +48,30 @@ func (h *unsafeOperationHandler) RemoveFailedStores(w http.ResponseWriter, r *ht
...
+		h.rd.JSON(w, http.StatusBadRequest, err)
 		return
 	}
-	storeSlice, ok := typeutil.JSONToUint64Slice(input["stores"])
-	if !ok {
-		h.rd.JSON(w, http.StatusBadRequest, "Store ids are invalid")
-		return
-	}
+
+	_, force := input["force"]
```
Comment (rleungx) [inline server/api/unsafe_operation.go:55]: How about using a query?

File: server/api/unsafe_operation.go [hunk 2]
```diff
@@ -48,24 +48,30 @@ func (h *unsafeOperationHandler) RemoveFailedStores(w http.ResponseWriter, r *ht
 	rc := getCluster(r)
 	var input map[string]interface{}
 	if err := apiutil.ReadJSONRespondError(h.rd, w, r.Body, &input); err != nil {
+		h.rd.JSON(w, http.StatusBadRequest, err)
```
Comment (rleungx) [inline server/api/unsafe_operation.go:51]:
```suggestion
		h.rd.JSON(w, http.StatusBadRequest, err.Error())
```

File: server/cluster/unsafe_recovery_controller.go [hunk 3]
```diff
@@ -164,21 +165,21 @@ func (u *unsafeRecoveryController) IsRunning() bool {
...
-func (u *unsafeRecoveryController) RemoveFailedStores(failedStores map[uint64]struct{}, timeout uint64) error {
+func (u *unsafeRecoveryController) RemoveFailedStores(failedStores map[uint64]struct{}, timeout uint64, force bool) error {
 	if u.IsRunning() {
 		return errs.ErrUnsafeRecoveryIsRunning.FastGenByArgs()
 	}
 	u.Lock()
 	defer u.Unlock()
 
-	if len(failedStores) == 0 {
+	if len(failedStores) == 0 && !force {
```
Comment (rleungx) [inline server/cluster/unsafe_recovery_controller.go:175]: I think we can skip the below logic if force is true.

File: server/cluster/unsafe_recovery_controller.go [hunk 4]
```diff
@@ -110,6 +110,7 @@ type unsafeRecoveryController struct {
 	step         uint64
 	failedStores map[uint64]struct{}
 	timeout      time.Time
+	force        bool
```
Comment (rleungx) [inline server/cluster/unsafe_recovery_controller.go:113]: Maybe we can choose a better name.

File: server/api/unsafe_operation.go (hunk 1 above)
Comment (Connor1996) [reply server/api/unsafe_operation.go:55]: what do you mean?

Comment (rleungx) [reply server/api/unsafe_operation.go:55]: Something like r.URL.Query().Get("force")

File: server/cluster/unsafe_recovery_controller.go [hunk 5]
```diff
@@ -615,13 +625,22 @@ func (u *unsafeRecoveryController) recordAffectedRegion(region *metapb.Region) {
 	}
 }
 
+func (u *unsafeRecoveryController) isFailed(peer *metapb.Peer) bool {
+	_, isFailed := u.failedStores[peer.StoreId]
+	_, isLive := u.storeReports[peer.StoreId]
```
Comment (rleungx) [inline server/cluster/unsafe_recovery_controller.go:630]: Do we need to wait for at least a round of store heartbeat?

Comment (Connor1996) [reply server/cluster/unsafe_recovery_controller.go:630]: can't avoid that for all the corner cases, adds caveat in the flag comment

File: server/api/unsafe_operation_test.go [hunk 6]
```diff
@@ -83,3 +84,23 @@ func (suite *unsafeOperationTestSuite) TestRemoveFailedStores() {
...
+
+	input = map[string]interface{}{"auto-detect": true}
+	data, _ = json.Marshal(input)
+	err = tu.CheckPostJSON(testDialClient, suite.urlPrefix+"/remove-failed-stores", data, tu.StatusOK(re))
+	suite.NoError(err)
+
+	// Test show
+	var output []cluster.StageOutput
+	err = tu.ReadGetJSON(re, testDialClient, suite.urlPrefix+"/remove-failed-stores/show", &output)
+	suite.NoError(err)
```
Comment (rleungx) [inline server/api/unsafe_operation_test.go:105]: Suggest checking the output, not just the error.

**Review by rleungx — COMMENTED** [review]
The rest LGTM.

File: server/api/unsafe_operation_test.go (hunk 6 above)
Comment (Connor1996) [reply server/api/unsafe_operation_test.go:105]: checking the output doesn't make any sense, the stage doesn't reach `finished` here

**Review by rleungx — APPROVED** [review]

**Review by disksing — APPROVED** [review]

**Pull request merged**